File: backend/room_manager.go
package main

import (
	"fmt"
	"log"
	"net/http"

	"github.com/gorilla/websocket"
)

func NewRoomManager() *RoomManager {
	return &RoomManager{
		rooms: map[string]*Room{},
		Upgrader: websocket.Upgrader{
			CheckOrigin: func(r *http.Request) bool { return true },
		},
	}
}

func (rm *RoomManager) findAvailableRoom() *Room {
	rm.mutex.Lock()
	defer rm.mutex.Unlock()

	log.Printf("Looking for available room. Total rooms: %d", len(rm.rooms))

	for _, room := range rm.rooms {
		room.mutex.Lock()
		count := 0
		for client := range room.clients {
			if client.registered {
				count++
			}
		}
		status := room.status
		room.mutex.Unlock()

		log.Printf("Room %s: players=%d, status=%v", room.id, count, status)

		if count < 4 && status == Waiting {
			log.Printf("Found available room: %s", room.id)
			return room
		}
	}

	room := rm.createRoom()
	rm.rooms[room.id] = room
	log.Printf("Created new room: %s", room.id)
	return room
}

func (rm *RoomManager) createRoom() *Room {
	roomID := fmt.Sprintf("room_%d", len(rm.rooms))
	room := &Room{
		clients:  make(map[*Client]bool),
		status:   Waiting,
		id:       roomID,
		messages: make([]Message, 0),
		gameMap:  generateMap(),
	}
	return room
}

File: backend/room_manager.js
import { Room } from './room.js';
import { generateMap } from './map_generator.js';

export class RoomManager {
    constructor() {
        this.rooms = new Map();
    }

    findAvailableRoom() {
        console.log(`Looking for available room. Total rooms: ${this.rooms.size}`);

        for (const [id, room] of this.rooms.entries()) {
            const count = room.getRegisteredPlayersCount();
            console.log(`Room ${id}: players=${count}, status=${room.status}`);

            if (count < 4 && room.status === 'waiting') {
                console.log(`Found available room: ${id}`);
                return room;
            }
        }

        const room = this.createRoom();
        this.rooms.set(room.id, room);
        console.log(`Created new room: ${room.id}`);
        return room;
    }

    createRoom() {
        const roomId = `room_${Math.random().toString(16).slice(2, 10)}`;
        return new Room(roomId, generateMap());
    }

    destroyRoom(roomId) {
        if (this.rooms.has(roomId)) {
            this.rooms.delete(roomId);
            console.log(`Room ${roomId} has been destroyed.`);
        }
    }
}